micropython/utarfile-write/utarfile/write.py

```diff
@@ -0,0 +1,135 @@
+        else:
+            self.f = fileobj
+        # Read through the existing file.
+        while self.next():
+            pass
+        # Position at start of end block.
+        self.f.seek(self.offset)
+    else:
+        raise ValueError("mode " + mode + " not supported.")
+
+
+def addfile(self, tarinfo, fileobj=None):
+    # Write the header: 100 bytes of name, 8 bytes of mode in octal...
+    buf = bytearray(_BLOCKSIZE)
+    name = tarinfo.name
+    size = tarinfo.size
+    if tarinfo.isdir():
+        size = 0
+        if not name.endswith("/"):
+            name += "/"
+    hdr = uctypes.struct(uctypes.addressof(buf), _TAR_HEADER, uctypes.LITTLE_ENDIAN)
+    _setstring(hdr.name, name, 100)
+    _setstring(hdr.mode, "%06o " % (tarinfo.mode & 0o7777), 7)
+    _setstring(hdr.uid, "%06o " % tarinfo.uid, 7)
+    _setstring(hdr.gid, "%06o " % tarinfo.gid, 7)
+    _setstring(hdr.size, "%011o " % size, 12)
+    _setstring(hdr.mtime, "%011o " % tarinfo.mtime, 12)
+    _setstring(hdr.typeflag, "5" if tarinfo.isdir() else "0", 1)
+    # Checksum is calculated with checksum field all blanks.
+    _setstring(hdr.chksum, " " * 8, 8)
+    # Calculate and insert the actual checksum.
+    chksum = sum(buf)
+    _setstring(hdr.chksum, "%06o\0" % chksum, 7)
+    # Emit the header.
+    self.f.write(buf)
+    self.offset += len(buf)
+
+    # Copy the file contents, if any.
+    if fileobj:
+        n_bytes = self.f.write(fileobj.read())
+        self.offset += n_bytes
+        remains = -n_bytes & (_BLOCKSIZE - 1)  # == 0b111111111
+        if remains:
+            buf = bytearray(remains)
+            self.f.write(buf)
+            self.offset += len(buf)
+
+
+def add(self, name, recursive=True):
+    # self.TarInfo will exist when this method is pasted into TarFile.
+    tarinfo = TarInfo(name)
+    try:
+        stat = os.stat(name)
+        tarinfo.mode = stat[0]
+        tarinfo.uid = stat[4]
+        tarinfo.gid = stat[5]
+        tarinfo.size = stat[6]
+        tarinfo.mtime = stat[8]
+    except OSError:
+        print("Cannot stat", name, " - skipping.")
+        return
+    if not (tarinfo.isdir() or tarinfo.isreg()):
+        # We only accept directories or regular files.
+        print(name, "is not a directory or regular file - skipping.")
+        return
+    if tarinfo.isdir():
+        self.addfile(tarinfo)
+        if recursive:
+            for f in os.ilistdir(name):
+                self.add(name + "/" + f[0], recursive)
+    else:  # type == REGTYPE
+        self.addfile(tarinfo, open(name, "rb"))
+
+
+def close(self):
+    # Must be called to complete writing a tar file.
+    if self.mode == "w":
+        self.f.write(_NUL * (_BLOCKSIZE * 2))
+        self.offset += _BLOCKSIZE * 2
+        remainder = self.offset % _RECORDSIZE
+        if remainder:
+            self.f.write(_NUL * (_RECORDSIZE - remainder))
+    self.f.close()
+
+
+# Inject extra functionality into TarFile.
+from . import TarFile, TarInfo
```

Comment:
This is a circular import, but I think it works because the package (`__init__`) is already loaded by this point.

Would it be smaller code to do the following within the `TarFile` class:
```
try:
    from .write import _open_write, addfile, add, close
except:
    pass
```
??